### PR TITLE
PIM-6645: Redesign of the sequential edit

### PR DESCRIPTION
**Pull request by juliensnz**
**Description (for Contributor and Core Developer)**

Redesign of the sequential edit. In this PR we remove all backend related stuff and create a simple SequentialEditController to handle the conversion from grid filters ton product ids. Se front didin't changed a lot as we don't have enough time to do it.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | NA
| Added Behats                      | NA
| Added integration tests           | NA
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

**Comment by BitOne** [comment]
Why are you using a specific cursor ? It was not possible to use the standard cursor and standard PQB by providing product ids ?

**Comment by juliensnz** [comment]
I needed to override it to select on ids instead of identifiers. WIth the current implementation, it's not possible to change that :/

File: src/Pim/Bundle/DataGridBundle/Normalizer/IdEncoder.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * Id encoder to manipulate product and product model ids
+ *
+ * @author    Julien Sanchez <julien@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class IdEncoder
+{
+    const PRODUCT_TYPE = 'product';
```
Comment (damien-carcel) [inline src/Pim/Bundle/DataGridBundle/Normalizer/IdEncoder.php]: public const

File: src/Pim/Bundle/DataGridBundle/Normalizer/IdEncoder.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
+<?php
```
Comment (damien-carcel) [inline src/Pim/Bundle/DataGridBundle/Normalizer/IdEncoder.php]: When creating new class with PHP 7.1 syntax, remember to always add
```
<?php

declare(strict_types=1);
```

File: src/Pim/Bundle/DataGridBundle/Normalizer/IdEncoder.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    /**
+     * Encode id and type to a type_id format.
+     *
+     * @param string $type
+     * @param string $id
+     *
+     * @return string
+     */
+    public static function encode($type, $id): string
```
Comment (damien-carcel) [inline src/Pim/Bundle/DataGridBundle/Normalizer/IdEncoder.php]: please typeint `$type` and `$id`

File: src/Pim/Bundle/DataGridBundle/Normalizer/IdEncoder.php (hunk 2 above)
Comment (juliensnz) [reply src/Pim/Bundle/DataGridBundle/Normalizer/IdEncoder.php]: done :)

**Comment by jjanvier** [comment]
@BitOne we discussed that with Juju.

We know all the PQB cursors and cursor factories are not the ideal solution, but we don't have the time right now to refactor it sadly.
The cool thing is that this PR drastically simplify the sequential edit even if it introduces a new cursor.

File: src/Pim/Bundle/DataGridBundle/Resources/config/mass_actions.yml [hunk 4]
```diff
@@ -60,6 +60,12 @@ services:
         tags:
             - { name: oro_datagrid.extension.mass_action.type, type: edit }
 
+    pim_datagrid.extension.mass_action.type.sequential_edit:
+        class: Pim\Bundle\DataGridBundle\Extension\MassAction\Actions\Redirect\SequentialEditMassAction
```
Comment (SamirBoulil) [inline src/Pim/Bundle/DataGridBundle/Resources/config/mass_actions.yml]: Don't you want to use a parameter here ?

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/SequentialEditController.php [hunk 5]
```diff
@@ -0,0 +1,85 @@
+<?php
+
```
Comment (SamirBoulil) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/SequentialEditController.php]: declare strict type and typehinting missing

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/SequentialEditController.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        $filters = $this->filterAdapter->adapt($parameters);
+
+        $products = [];
+        $cursor = $this->getProductsCursor($filters, [
+            'locale' => $parameters['dataLocale'],
+            'scope'  => $parameters['dataScope'],
+            'sort'   => $parameters['sort']
+        ]);
+
+        while ($cursor->valid() && $cursor->key() < 1000) {
```
Comment (SamirBoulil) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/SequentialEditController.php]: Can you put 1000 in a constant so we know what it stands for ?

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/SequentialEditController.php (hunk 5 above)
Comment (SamirBoulil) [reply src/Pim/Bundle/EnrichBundle/Controller/Rest/SequentialEditController.php]: ah it was a move ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/provider/sequential-edit-provider.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
+'use strict';
+/**
+ * Sequential edit
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/provider/sequential-edit-provider.js]: Maybe it misses some explanations on what it is

File: src/Pim/Bundle/DataGridBundle/Resources/config/mass_actions.yml (hunk 4 above)
Comment (juliensnz) [reply src/Pim/Bundle/DataGridBundle/Resources/config/mass_actions.yml]: It was not the case for other actions so I preferred to stay consistent

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/SequentialEditController.php (hunk 6 above)
Comment (juliensnz) [reply src/Pim/Bundle/EnrichBundle/Controller/Rest/SequentialEditController.php]: done

**Comment by juliensnz** [comment]
@jjanvier Will do for the changelog and I changed a behat to do what you are asking ofr :)

**Pull request merged**